src/stmt.cpp

```diff
@@ -1429,17 +1429,22 @@ void ForeachStmt::EmitCode(FunctionEmitContext *ctx) const {
         startVals.push_back(sv);
         endVals.push_back(ev);
 
+        const Type *svTy = startExprs[i]->GetType();
+        const Type *evTy = endExprs[i]->GetType();
+
         // nItems = endVal - startVal
-        llvm::Value *nItems = ctx->BinaryOperator(llvm::Instruction::Sub, ev, sv, "nitems");
+        llvm::Value *nItems = ctx->BinaryOperator(llvm::Instruction::Sub, ev, sv,
+                                                  !svTy->IsUnsignedType() && !evTy->IsUnsignedType(), "nitems");
```

Comment:
From the documentation: "each dimension is specified by identifier = start ... end, where start is a signed integer value less than or equal to end". I.e. start and end ranges should be treated as signed integers.

Also, it make sense to cover this with tests.

Reply:
Ok, the source code is fixed, but the test is not yet added.